modules/storages/app/common/storages/peripherals/storage_interaction/authentication_strategies/sso_user_token.rb

```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal:true
+
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module Storages
+  module Peripherals
+    module StorageInteraction
+      module AuthenticationStrategies
+        class SsoUserToken
+          def self.strategy
+            Strategy.new(:sso_user_token)
+          end
+
+          def initialize(user)
+            @user = user
+          end
+
+          def call(storage:, http_options: {}, &)
+            # README: important remark: this code will NOT automatically refresh, if the token does not have
+            # information about the expiration date. The question is, IF we want to support this case.
+            # In the case above, the token will be used without refresh, which will lead to an unauthorized response
+            # once the token is no longer valid. This can only be solved right now by logging out
+            # and in again - which will replace the account token itself.
+            # Open question: Do we want to support this case?
```

Comment:
My opinion: If we want to keep our implementation simple here by enforcing that we know an expiration time, we should at least try to carry the expiration time communicated by the token endpoint along. This should give us an expiration time for the majority of OAuth 2.0 implementations.

If you agree I'd write a feature for that and would probably exchange with Wieland whether this limitation is acceptable.

Reply:
Let me back up: Right now we only know of an expiration time if we can parse the token as JWT, read it successfully and extract the `exp` claim. In all other situations, we don't know the expiration and would try to use the token forever.

You are suggesting here, that we never try to refresh the token, unless we know that it's expired.

My opinion: Right now, this is not good enough, because we would fail too often.

My suggestion: We should know the expiration time in more cases, to make your proposal feasible. That's the feature I want to write. Important remark is that we will not be able to cover 100% of cases according to the OAuth 2.0 specification. That's where I'd clarify with Wieland, that we are "good enough" from a product perspective.

Reply:
You have my support in raising this as a feature proposal ;)